app/uk/gov/hmrc/eoricommoncomponent/frontend/CdsErrorHandler.scala

```diff
@@ -69,6 +69,11 @@ class CdsErrorHandler @Inject() (
         logger.info("Session time out: " + sessionTimeOut.errorMessage, exception)
         // $COVERAGE-ON
         Future.successful(Redirect(SecuritySignOutController.displayPage(service)).withNewSession)
+      case illegalStateException: IllegalStateException =>
```

Comment:
I think the ask was misunderstood. This exception is thrown in a very specific circumstance. For one, we shouldn't be so liberally throwing exceptions, it's not good practice and is essentially abandoning any control over the programs flow.

This is explicitly stated in the ticket.

The intention was that we go to where we make the check in the session cache for the user flow, that we handle it as described in the ticket. Not simply to change the error handler for unhandled exceptions.

You should be looking at RequestSessionData. Probably the userSubscriptionFlow def. It shouldn't throw an exception, instead we could do an Either[SessionError, SubscriptionFlow] so that if there is no subscription flow we can return the 'SessionError' (we'll need to define this) and handle it in the controller.

Reply:
Some examples

https://github.com/hmrc/common-transit-convention-traders/blob/main/app/v2/models/errors/PersistenceError.scala - for SessionError (i.e. a base trait and then some extensions that we'd actually return)

and an example of how I'd expect something like this to ideally be handled
https://github.com/hmrc/common-transit-convention-traders/blob/03e3d4148150f2718fcc5389bd1c5382505fd058/app/v2/services/PersistenceService.scala#L87

Note how we have the recover block (because there are underlying exceptions that get thrown that we don't control) but we don't just rethrow the exception, instead we return a Left that carries the error so we can check upstream.

If you want we can go over this pattern.